azkaban-common/src/main/java/azkaban/utils/RestfulApiClient.java

```diff
@@ -76,74 +73,25 @@ public static URI buildUri(final String host, final int port, final String path,
       }
     }
 
-    URI uri = null;
     try {
-      uri = builder.build();
+      return builder.build();
     } catch (final URISyntaxException e) {
       throw new IOException(e);
     }
-
-    return uri;
-  }
-
-  /**
-   * helper function to build a valid URI.
-   *
-   * @param uri the URI to start with.
-   * @param params extra query parameters to append.
-   * @return the URI built from the inputs.
-   */
-  public static URI BuildUri(final URI uri, final Pair<String, String>... params)
-      throws IOException {
-    final URIBuilder builder = new URIBuilder(uri);
-
-    if (params != null) {
-      for (final Pair<String, String> pair : params) {
-        builder.setParameter(pair.getFirst(), pair.getSecond());
-      }
-    }
-
-    URI returningUri = null;
-    try {
-      returningUri = builder.build();
-    } catch (final URISyntaxException e) {
-      throw new IOException(e);
-    }
-
-    return returningUri;
-  }
-
-  /**
-   * helper function to fill  the request with header entries .
-   */
-  private static HttpMessage completeRequest(final HttpMessage request,
-      final List<NameValuePair> headerEntries) {
-    if (null == request) {
-      logger.error("unable to complete request as the passed request object is null");
-      return request;
-    }
-
-    // dump all the header entries to the request.
-    if (null != headerEntries && headerEntries.size() > 0) {
-      for (final NameValuePair pair : headerEntries) {
-        request.addHeader(pair.getName(), pair.getValue());
-      }
-    }
-    return request;
   }
 
   /**
    * helper function to fill  the request with header entries and posting body .
    */
   private static HttpEntityEnclosingRequestBase completeRequest(
       final HttpEntityEnclosingRequestBase request,
-      final List<NameValuePair> headerEntries,
-      final String postingBody) throws UnsupportedEncodingException {
-    if (null != completeRequest(request, headerEntries)) {
-      // dump the post body UTF-8 will be used as the default encoding type.
-      if (null != postingBody && postingBody.length() > 0) {
-        final HttpEntity entity = new ByteArrayEntity(postingBody.getBytes("UTF-8"));
-        request.setHeader("Content-Length", Long.toString(entity.getContentLength()));
+      final List<Pair<String, String>> params) throws UnsupportedEncodingException {
+    if (request != null) {
+      if (null != params && !params.isEmpty()) {
+        final List<NameValuePair> formParams = params.stream()
+            .map(pair -> new BasicNameValuePair(pair.getFirst(), pair.getSecond()))
+            .collect(Collectors.toList());
+        final HttpEntity entity = new UrlEncodedFormEntity(formParams, "UTF-8");
```

Comment:
In real usage `UrlEncodedFormEntity` provides the `"Content-Length"` header. Request becomes invalid if a duplicate header is also added manually. The unit test overrides `sendAndReturn()` so the `"Content-Length"` header is not seen in the unit test any more.